src/H5Dselect.c

```diff
@@ -256,6 +256,187 @@ H5D__select_io(const H5D_io_info_t *io_info, size_t elmt_size, size_t nelmts, co
     FUNC_LEAVE_NOAPI(ret_value)
 } /* end H5D__select_io() */
 
+/*-------------------------------------------------------------------------
+ * Function:    H5D_select_io_mem
+ *
+ * Purpose:     Perform memory copies directly between two memory buffers
+ *              according to the selections in the `dst_space` and
+ *              `src_space` dataspaces.
+ *
+ * Note:        This routine is [basically] the same as H5D__select_io,
+ *              with the only difference being that the readvv/writevv
+ *              calls are exchanged for H5VM_memcpyvv calls. Changes should
+ *              be made to both routines.
+ *
+ * Return:      Non-negative on success/Negative on failure
+ *
+ *-------------------------------------------------------------------------
+ */
+herr_t
+H5D_select_io_mem(void *dst_buf, const H5S_t *dst_space, const void *src_buf, const H5S_t *src_space,
+                  size_t elmt_size, size_t nelmts)
+{
+    H5S_sel_iter_t *dst_sel_iter      = NULL;  /* Destination dataspace iteration info */
+    H5S_sel_iter_t *src_sel_iter      = NULL;  /* Source dataspace iteration info */
+    hbool_t         dst_sel_iter_init = FALSE; /* Destination dataspace selection iterator initialized? */
+    hbool_t         src_sel_iter_init = FALSE; /* Source dataspace selection iterator initialized? */
+    hsize_t *       dst_off           = NULL;  /* Pointer to sequence offsets in destination buffer */
+    hsize_t *       src_off           = NULL;  /* Pointer to sequence offsets in source buffer */
+    size_t *        dst_len           = NULL;  /* Pointer to sequence lengths in destination buffer */
+    size_t *        src_len           = NULL;  /* Pointer to sequence lengths in source buffer */
+    size_t          curr_dst_seq;              /* Current destination buffer sequence to operate on */
+    size_t          curr_src_seq;              /* Current source buffer sequence to operate on */
+    size_t          dst_nseq;                  /* Number of sequences generated for destination buffer */
+    size_t          src_nseq;                  /* Number of sequences generated for source buffer */
+    size_t          dxpl_vec_size;             /* Vector length from API context's DXPL */
+    size_t          vec_size;                  /* Vector length */
+    ssize_t         bytes_copied;
```

Comment:
Not critical for this PR, but we should consider refactoring H5VM_memcpyvv to return herr_t and return the number of bytes copied via an out parameter.

Reply:
Agreed. A memcpy implementation is certainly one where I could see the range of ssize_t being a problem one day. I'll make a note on changing this.